### PR TITLE
[wasm] Add Timezone data generator as build task

### PR DESCRIPTION
**Pull request by tqiu8**

### CODE REVIEW

File: src/mono/wasm/wasm.targets [hunk 1]
```diff
@@ -32,9 +34,20 @@
                            OutputPath="$(WasmPInvokeTablePath)" />
   </Target>
 
+  <Target Name="BundleWasm" DependsOnTargets="CheckEnv">
```
Comment (safern) [inline src/mono/wasm/wasm.targets:37]: Nit: this target is bundling the timezone, should we rename the target, or are we expecting to add more calls to this task with a different bundle `type`?

File: src/mono/wasm/wasm.targets [hunk 2]
```diff
@@ -32,9 +34,20 @@
...
+    <PropertyGroup>
+      <WasmBundleTaskPath>$(NativeBinDir)dotnet.timezones.blat</WasmBundleTaskPath>
```
Comment (safern) [inline src/mono/wasm/wasm.targets:39]:
```suggestion
      <TimeZonesDataPath>$(NativeBinDir)dotnet.timezones.blat</TimeZonesDataPath>
```

File: src/mono/wasm/wasm.targets [hunk 3]
```diff
@@ -32,9 +34,20 @@
...
+    </PropertyGroup>
+    <WasmBundleTask
+      InputDirectory="$(ArtifactsObjDir)wasm/data/input"
+      OutputDirectory="$(ArtifactsObjDir)wasm/data/output"
+      Type="timezone"
+      FileName="$(WasmBundleTaskPath)" />
```
Comment (safern) [inline src/mono/wasm/wasm.targets:45]:
```suggestion
      FileName="$(TimeZonesDataPath)" />
```

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        };
+
+        var jsonBytes = JsonSerializer.SerializeToUtf8Bytes(indices);
+        
+        return (jsonBytes, stream);
+    }
+
+    private (byte[] json_bytes, MemoryStream stream) readTimeZone (string folder) {
+        // https://en.wikipedia.org/wiki/Tz_database#Area
+        var areas = new[] { "Africa", "America", "Antarctica", "Arctic", "Asia", "Atlantic", "Australia", "Europe", "Indian", "Pacific", "zone1970.tab"};
```
Comment (safern) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:76]: Should these areas be accepted as an input by the task so that in the future is controllable from outside of it via a MSBuild item?

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+        return enumerateData (areas, folder);
+    }
+
+    private (byte[] json_bytes, MemoryStream stream) readGeneralData (string input_folder) {
+        var DirectoryInfo = new DirectoryInfo (input_folder);
+        string[] sub_folders = DirectoryInfo.EnumerateFileSystemInfos().Select(f => f.Name).ToArray();
+
+        return enumerateData (sub_folders, input_folder);
```
Comment (safern) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:85]: nit: 
```suggestion
        return EnumerateData(sub_folders, input_folder);
```

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        var jsonBytes = JsonSerializer.SerializeToUtf8Bytes(indices);
+        
+        return (jsonBytes, stream);
+    }
+
+    private (byte[] json_bytes, MemoryStream stream) readTimeZone (string folder) {
+        // https://en.wikipedia.org/wiki/Tz_database#Area
+        var areas = new[] { "Africa", "America", "Antarctica", "Arctic", "Asia", "Atlantic", "Australia", "Europe", "Indian", "Pacific", "zone1970.tab"};
+
+        return enumerateData (areas, folder);
```
Comment (safern) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:78]: Nit:
```suggestion
        return EnumerateData (areas, folder);
```

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+using System.Diagnostics;
+
+public class WasmBundleTask : Task
+{
+    public string? InputDirectory { get; set; }
+    public string? OutputDirectory { get; set; }
+    public string? Type { get; set; }
+    public string? FileName { get; set; } 
+
+    private (byte[] json_bytes, MemoryStream stream) enumerateData (string[] sub_folders, string output_folder) {
```
Comment (safern) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:22]:
```suggestion
    private (byte[] json_bytes, MemoryStream stream) EnumerateData (string[] sub_folders, string output_folder) {
```

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        {
+            WriteIndented = false
+        };
+
+        var jsonBytes = JsonSerializer.SerializeToUtf8Bytes(indices);
+        
+        return (jsonBytes, stream);
+    }
+
+    private (byte[] json_bytes, MemoryStream stream) readTimeZone (string folder) {
```
Comment (safern) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:74]:
```suggestion
    private (byte[] json_bytes, MemoryStream stream) ReadTimeZone (string folder) {
```

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        // https://en.wikipedia.org/wiki/Tz_database#Area
+        var areas = new[] { "Africa", "America", "Antarctica", "Arctic", "Asia", "Atlantic", "Australia", "Europe", "Indian", "Pacific", "zone1970.tab"};
+
+        return enumerateData (areas, folder);
+    }
+
+    private (byte[] json_bytes, MemoryStream stream) readGeneralData (string input_folder) {
```
Comment (safern) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:81]:
```suggestion
    private (byte[] json_bytes, MemoryStream stream) ReadGeneralData (string input_folder) {
```

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        var DirectoryInfo = new DirectoryInfo (input_folder);
```
Comment (safern) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:82]: Usually in the managed code in runtime we don't have a space in between the method name and the open parenthesis.

File: src/mono/wasm/wasm.targets (hunk 1 above)
Comment (tqiu8) [reply src/mono/wasm/wasm.targets:37]: I believe we eventually want this task to be used with other data types besides just the TimeZone data.

**Review by lewing — COMMENTED** [review]
This is a good start.  Would it make sense to make two tasks?  One that only made a bundle and that handled the timezone specific parts and to use as input to the bundle?  I do think it would be useful expose the bundling logic for a general set of targets.

**Comment by tqiu8** [comment]
> This is a good start. Would it make sense to make two tasks? One that only made a bundle and that handled the timezone specific parts and to use as input to the bundle? I do think it would be useful expose the bundling logic for a general set of targets.

Yes, I can separate this task into a TimeZone specific one that handles the downloading, compiling, and reordering of data; and another task that handles general input and bundles into a data file.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        foreach (var folder in sub_folders)
+        {
+            var path = Path.Combine(output_folder, folder);
+            if (folder == "zone1970.tab") {
+                var fileInfo = new FileInfo(path);
+                indices.Add(new object[] { "zone.tab", fileInfo.Length});
+                string[] systemtz = { "America/Los_Angeles", "Australia/Sydney", "Europe/London", "Pacific/Tongatapu", 
+                                "America/Sao_Paulo", "Australia/Perth", "Africa/Nairobi", "Europe/Berlin",
+                                "Europe/Moscow", "Africa/Tripoli", "America/Argentina/Catamarca", "Europe/Lisbon",
+                                "America/St_Johns"};
```
[the remark this answers is not in the document]
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:35]: These are just the timezones that will be cached for SystemTimeZones. The original file was 19kb so we filtered it down to those timezones.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        var DirectoryInfo = new DirectoryInfo (input_folder);
+        string[] sub_folders = DirectoryInfo.EnumerateFileSystemInfos().Select(f => f.Name).ToArray();
+
+        return enumerateData (sub_folders, input_folder);
+    }
+
+    private void DownloadTimeZoneData (string input_folder, string output_folder) {
+        using (var client = new WebClient())
+        {
+            client.DownloadFile("https://data.iana.org/time-zones/tzdata-latest.tar.gz", $"{input_folder}/tzdata.tar.gz");
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:91]: We cannot depend on latest data automatically. It's be better to move the data download outside of this task and set the version explicitly there

**Review by marek-safar — CHANGES_REQUESTED** [review]
.

**Review by lewing — CHANGES_REQUESTED** [review]
Great work! The bundler split looks good.  For the versioning I think we can make things a little simpler still.  The timezone data exists at `https://data.iana.org/time-zones/tzdb<version>/` already unpacked so I think it makes sense to make the TimeZoneLoader task take a version property like `2020a` then instead of downloading the zip you can download the individual files from `https://data.iana.org/time-zones/tzdb2020a/<path>` and build the bundle from that.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+using System.Net;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+
+public class WasmBundleTask : Task
+{
+    [Required]
+    public string? AppDir { get; set; }
```
Comment (lewing) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:19]: This is unused?

Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:19]: Oops yes let me take that out.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    public string? InputDirectory { get; set; }
+    public string? FileName { get; set; } 
+
+    private (byte[] json_bytes, MemoryStream stream) EnumerateData() {
+        var indices = new List<object[]>();
+        var stream = new MemoryStream();
+        
+        var directoryInfo = new DirectoryInfo(InputDirectory!);
+        
+        foreach (var entry in directoryInfo.EnumerateFiles("*", SearchOption.AllDirectories)) {
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:27]: Maybe `s,var entry, var fileInfo,`

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+        data = EnumerateData();
+        
+        if (FileName == null) {
+            throw new ArgumentException($"Invalid file name");
+        }
+        using (var file = File.Open(FileName, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.ReadWrite))
+        {
+            var jsonBytes = data.json_bytes;
+            var stream = data.stream;
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:58]: these seem unnecessary, you can directly use `data.json_bytes`, and `data.stream`.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+            var bytes = new byte[4];
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:59]: Maybe rename to `lengthBytes`, or something like that?

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+using System.Text.Json.Serialization;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Net;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+
+public class WasmBundleTask : Task
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:16]: Maybe rename to `GenerateWasmBundle`, or `CreateWasmBundle`

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj [hunk 18]
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <OutputType>Library</OutputType>
+    <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj:5]: Why is this needed? If this can be dropped, then the `<Compile..` items can be removed too.

File: eng/testing/tests.mobile.targets [hunk 19]
```diff
@@ -162,6 +162,23 @@
       AssemblySearchPaths="@(AssemblySearchPaths)" />
   </Target>
 
+  <UsingTask TaskName="TimeZoneLoader"
+             AssemblyFile="$(WasmBundleTaskTasksAssemblyPath)" />
+  <Target Condition="'$(TargetOS)' == 'Browser'" Name="LoadTimeZone" >
```
Comment (radical) [inline eng/testing/tests.mobile.targets:167]: Place the name before the condition.
And maybe rename the target to `LoadTimeZoneData`?

File: eng/testing/tests.mobile.targets [hunk 20]
```diff
@@ -162,6 +162,23 @@
...
+    <TimeZoneLoader
+      InputDirectory="$(BundleDir)obj/data/input"
+      OutputDirectory="$(BundleDir)obj/data/output"
+      Version="2020a"/>
```
Comment (radical) [inline eng/testing/tests.mobile.targets:171]: Maybe this should be made a property? `$(TimeZoneDataVersion)`? Or whatever this version is.

File: eng/testing/tests.mobile.targets [hunk 21]
```diff
@@ -162,6 +162,23 @@
...
+  </Target>
+
+  <UsingTask TaskName="WasmBundleTask"
+             AssemblyFile="$(WasmBundleTaskTasksAssemblyPath)" />
+  <Target Condition="'$(TargetOS)' == 'Browser'" Name="BundleWasmTestData" DependsOnTargets="LoadTimeZone">
```
Comment (radical) [inline eng/testing/tests.mobile.targets:176]:
```xml
<Target
  Name="BundleWasmTestData"  
  Condition="'$(TargetOS)' == 'Browser'"
  DependsOnTargets="LoadTimeZone">
```

File: src/mono/wasm/wasm.targets [hunk 22]
```diff
@@ -35,10 +35,28 @@
                            OutputPath="$(WasmPInvokeTablePath)" />
   </Target>
 
+  <UsingTask TaskName="TimeZoneLoader" AssemblyFile="$([MSBuild]::NormalizePath('$(ArtifactsBinDir)', 'WasmBundleTask', 'Debug', '$(NetCoreAppCurrent)', 'publish', 'WasmBundleTask.dll'))"/>
+  <Target Name="LoadTimeZones" DependsOnTargets="CheckEnv">
+  <TimeZoneLoader
```
Comment (radical) [inline src/mono/wasm/wasm.targets:40]: indentation

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+{
+    public string? InputDirectory { get; set; }
+    public string? FileName { get; set; } 
+
+    private (byte[] json_bytes, MemoryStream stream) EnumerateData() {
+        var indices = new List<object[]>();
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:22]: Maybe `var fileTable`?

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Build" Version="$(RefOnlyMicrosoftBuildVersion)" />
+    <PackageReference Include="Microsoft.Build.Framework" Version="$(RefOnlyMicrosoftBuildFrameworkVersion)" />
+    <PackageReference Include="Microsoft.Build.Tasks.Core" Version="$(RefOnlyMicrosoftBuildTasksCoreVersion)" />
+    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(RefOnlyMicrosoftBuildUtilitiesCoreVersion)" />
+    <PackageReference Include="System.Reflection.MetadataLoadContext" Version="4.7.1" />
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj:13]: Is this needed?

File: tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+using System.Reflection;
+using System.Diagnostics;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class TimeZoneLoader : Task
+{
+    public string? InputDirectory { get; set; }
+    public string? OutputDirectory { get; set; }
+    public string? Version { get; set; }
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs:19]: Add `[Required]`

File: tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+                process.Start();
+                process.WaitForExit();
+            }
+        }
+        File.Copy(Path.Combine(InputDirectory!,"zone1970.tab"), Path.Combine(OutputDirectory!,"zone1970.tab"));
+    }
+
+    private void FilterTimeZoneData(string[] areas) {
+        var directoryInfo = new DirectoryInfo (OutputDirectory!);
+        foreach (var entry in directoryInfo.EnumerateDirectories()) {
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs:46]: Can we use `foreach (var info in Directory.EnumerateDirectories(OutputDirectory))` here?

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+        
+        return (jsonBytes, stream);
+    }
+
+    public override bool Execute()
+    {
+        (byte[] json_bytes, MemoryStream stream) data;
+
+        if (InputDirectory == null) {
+            throw new ArgumentException("Input directory doesn't exist.");
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:47]: Also, instead of throwing exceptions for invalid input, use:

```csharp
Log.LogError($"Input directory '{InputDirectory}' does not exist");
return false;
```

File: tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+using System.Text.Json.Serialization;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Net;
+using System.Reflection;
+using System.Diagnostics;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class TimeZoneLoader : Task
```
[the remark this answers is not in the document]
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs:15]: It seems like the other build tasks are named liked `WasmAppBuilder` or `PInvokeTableGenerator`

[the remark this answers is not in the document]
Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs:15]: The mobile.tasks do seem to be named like that, but not others like in `installer.tasks`. IMHO, the task names should read like actions, that get executed in a target. Not a blocker though.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs [hunk 29]
```diff
@@ -14,70 +14,80 @@
 
 public class TimeZoneLoader : Task
 {
+    [Required]
     public string? InputDirectory { get; set; }
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs:18]: The required ones shouldn't be nullable.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj (hunk 24 above)
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj:13]: System.Reflection.MetadataLoadContext is not needed, but the other references are for the required attribute to work.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs (hunk 29 above)
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs:18]: They need to be nullable if they're uninitialized: 
`TimeZoneLoader.cs(18,19): error CS8618: Non-nullable property 'InputDirectory' is uninitialized. Consider declaring the property as nullable. [/Users/tammyqiu/Desktop/Desktop/runtime/tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj]`

File: eng/testing/tests.mobile.targets (hunk 19 above)
Comment (tqiu8) [reply eng/testing/tests.mobile.targets:167]: Just following the ordering of other targets.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs (hunk 27 above)
Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:52]: Since, this is a Required input now, you don't need to check for null. Instead check for `Directory.Exists(InputDirectory)`, and `LogError`, if it doesn't.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+        (byte[] json_bytes, MemoryStream stream) data;
+
+        if (InputDirectory == null) 
+        {
+            throw new ArgumentException("Input directory doesn't exist.");
+        }
+
+        data = EnumerateData();
+        
+        if (FileName == null)
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:57]: Same idea here, since `FileName` is Required now, msbuild will ensure that it is not null. So, here do:

```
if (!File.Exists(FileName))
{
   Log.LogError($"Input filename '${FileName}' does not exist.");
   return false;
}
```

Logging the error like this ensures that a neat error message gets surfaced to the list of errors, and to the IDE.
And returning false here indicates to msbuild that the task failed.
Throwing an exception also does the same thing, except that you get exception trace, and message etc, which should happen only in case of something we didn't expect.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs (hunk 16 above)
Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:63]: Renaming like this just makes it clearer, and consistent with the other variable name `magicBytes`, making the code easier to read, and easier to spot mistakes.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj (hunk 24 above)
Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.csproj:13]: Yep, I was talking about the `MetadataLoadContext` in particular. The others would be required for all the msbuild bits, like `Task` etc. 👍

File: tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+using System.Net;
+using System.Reflection;
+using System.Diagnostics;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class TimeZoneLoader : Task
+{
+    [Required]
+    public string? InputDirectory { get; set; }
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs:18]: Why is this parameter needed at all? IIUC, this just seems to be the directory where files are temporarily downloaded. Do they get used after this task is done? If not, then I would suggest removing this, and instead creating a temporary directory, and downloading to that. And removing it at the end of the task.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/TimeZoneLoader.cs:15]: Since this is a new task that we are creating, we can follow the msbuild conventions. The idea here is that a msbuild Target executes a bunch of Tasks, which are actions/code, so, you would name the Tasks like a method.
Let's change the names for the new tasks to match that.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+        var jsonBytes = JsonSerializer.SerializeToUtf8Bytes(indices);
+        
+        return (jsonBytes, stream);
+    }
+
+    public override bool Execute()
+    {
+        (byte[] json_bytes, MemoryStream stream) data;
+
+        if (InputDirectory == null)
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:49]: Since, this is a Required input now, you don't need to check for null. Instead check for Directory.Exists(InputDirectory), and LogError, if it doesn't.

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+    public override bool Execute()
+    {
+        (byte[] json_bytes, MemoryStream stream) data;
+
+        if (InputDirectory == null) 
+        {
+            Log.LogError($"Input directory '{InputDirectory}' does not exist");
+            return false;
+        }
+
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:54]: Marking a property with `[Required]` attribute means that msbuild will ensure that it will have *some* value. It could be valid, or complete gibberish, but it won't be null.

So, we still need to check that the `FileName` property is a valid filename that exists.
```
if (!File.Exists(FileName))
{
   Log.LogError($"Input filename '${FileName}' does not exist.");
   return false;
}
```

Logging the error like this ensures that a neat error message gets surfaced to the list of errors, and to the IDE.
And returning false here indicates to msbuild that the task failed.

Throwing an exception also does the "same" thing, except that you get exception trace, and message etc, which should happen only in case of something we didn't expect.

**Review by radical — CHANGES_REQUESTED** [review]

File: tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs (hunk 33 above)
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:54]: the file is generated, so it wouldn't exist at the time.

Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBundleTask/WasmBundleTask.cs:54]: Ah! Ok, then we should name it as `OutputFileName`, to make it clearer.

**Review by lewing — APPROVED** [review]
This looks almost ready.  The remaining issues are mostly around naming (I do think OutputFileName is clearer).

**Review by radical — APPROVED** [review]
LGTM! Good job, and thanks for being patient with the review feedback!

File: tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs [hunk 34]
```diff
@@ -0,0 +1,102 @@
...
+    public string? Version { get; set; }
+
+    private void DownloadTimeZoneDataSource() 
+    {
+        List<string> files = new List<string>() {"africa", "antarctica", "asia", "australasia", "etcetera", "europe", "northamerica", "southamerica", "zone1970.tab"};
+        using (var client = new WebClient())
+        {
+            foreach (var file in files) 
+            {
+                client.DownloadFile($"https://data.iana.org/time-zones/tzdb-{Version}/{file}", $"{Path.Combine(InputDirectory!, file)}");
```
Comment (marek-safar) [inline tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs:33]: It might be helpful to print what it's doing in case the download takes long time

File: tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs [hunk 35]
```diff
@@ -0,0 +1,102 @@
...
+            {
+                process.StartInfo.Arguments = $"-d \"{OutputDirectory}\" \"{Path.Combine(InputDirectory!, f)}\"";
+                process.Start();
+                process.WaitForExit();
+            }
+        }
+        File.Copy(Path.Combine(InputDirectory!,"zone1970.tab"), Path.Combine(OutputDirectory!,"zone1970.tab"));
+    }
+
+    private void FilterTimeZoneData()
```
Comment (marek-safar) [inline tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs:53]: Why is this needed? You process additional files and then they are deleted?

**Review by marek-safar — APPROVED** [review]
Few more comments, otherwise LGTM

File: tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs (hunk 35 above)
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs:53]: Yes, this is just to remove extraneous files so that `CreateWasmBundle` doesn't end up including them in the data archive.

File: src/mono/wasm/wasm.targets [hunk 36]
```diff
@@ -35,10 +35,28 @@
                            OutputPath="$(WasmPInvokeTablePath)" />
   </Target>
 
+  <UsingTask TaskName="DownloadTimeZoneData" AssemblyFile="$([MSBuild]::NormalizePath('$(ArtifactsBinDir)', 'CreateWasmBundle', 'Debug', '$(NetCoreAppCurrent)', 'publish', 'CreateWasmBundle.dll'))"/>
```
[the remark this answers is not in the document]
Comment (tqiu8) [reply src/mono/wasm/wasm.targets:38]: hm I changed this to `$(Configuration)` and it throws:
```
/home/taqiu/runtime/src/mono/wasm/wasm.targets(40,5): error MSB4062: The "DownloadTimeZoneData" task could not 
be loaded from the assembly 
/home/taqiu/runtime/artifacts/bin/CreateWasmBundle/Release/net5.0/publish/CreateWasmBundle.dll. Could not load file 
or assembly '/home/taqiu/runtime/artifacts/bin/CreateWasmBundle/Release/net5.0/publish/CreateWasmBundle.dll'. The 
system cannot find the file specified. [/home/taqiu/runtime/src/libraries/src.proj]                            
/home/taqiu/runtime/src/mono/wasm/wasm.targets(40,5): error MSB4062:  Confirm that the <UsingTask> declaration is 
correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements 
Microsoft.Build.Framework.ITask. [/home/taqiu/runtime/src/libraries/src.proj]  
```

I was following the `<UsingTask TaskName="PInvokeTableGenerator" AssemblyFile="$([MSBuild]::NormalizePath('$(ArtifactsBinDir)', 'WasmAppBuilder', 'Debug', '$(NetCoreAppCurrent)', 'publish', 'WasmAppBuilder.dll'))"/>`

File: tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs (hunk 35 above)
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs:53]: This method removes unnecessary files so that the subsequent `CreateWasmBundle` does not include them in the data archive.

File: tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs [hunk 37]
```diff
@@ -0,0 +1,103 @@
...
+
+    [Required]
+    public string? Version { get; set; }
+
+    private void DownloadTimeZoneDataSource() 
+    {
+        List<string> files = new List<string>() {"africa", "antarctica", "asia", "australasia", "etcetera", "europe", "northamerica", "southamerica", "zone1970.tab"};
+        using (var client = new WebClient())
+        {
+            Console.WriteLine("Downloading TimeZone data files");
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/CreateWasmBundle/DownloadTimeZoneData.cs:31]: @tqiu8  Use `Log.LogMessageFromText ("Downloading TimeZone data files", MessageImportance.Low);` instead of CWL. This way, the message would end up in the logs correctly. Could you fix this as a follow up?
